components/suggest/src/store.rs

```diff
@@ -197,15 +235,23 @@ where
                         Some(icon_id) => dao.drop_icon(icon_id)?,
                         None => dao.drop_suggestions(&record_id)?,
                     };
+                    // Remove this `record.id` from unknowns.
                     dao.put_meta(LAST_INGEST_META_KEY, record.last_modified)?;
                     Ok(())
                 })?;
                 continue;
             }
             let Ok(fields) = serde_json::from_value(serde_json::Value::Object(record.fields.clone())) else {
                 // We don't recognize this record's type, so we don't know how
-                // to ingest its suggestions. Skip to the next record.
-                writer.write(|dao| dao.put_meta(LAST_INGEST_META_KEY, record.last_modified))?;
+                // to ingest its suggestions. Record this in the meta table.
+                writer.write(|dao| {
+                    // TODO: Read `unknowns`, update JSON, and write it back
+                    dao.put_meta(LAST_INGEST_META_KEY, record.last_modified)?;
+                    let mut unknown_records = dao.get_meta::<UnparsableRecords>(UNKNOWN_RECORDS_META_KEY)?.unwrap_or_default();
+                    unknown_records.0.insert(record.id.clone(), UnparsableRecord{schema_version: VERSION});
+                    dao.put_meta(UNKNOWN_RECORDS_META_KEY, unknown_records)?;
```

Comment:
Potential concern: this would load/store `UnparsableRecords` from the meta table for every single unknown record observed. What do you think about stashing all the unknown records and storing them in bulk at the end of the ingestion?

Reply:
Tif actually did it that way at first, and we chatted about the different approaches over Zoom! 😁

Storing the unknown record IDs in bulk at the end of the ingestion avoids the extra loads and stores—but risks losing the IDs if we're interrupted during ingestion. Interruptions might be for reasons outside of our control, like the OS terminating the background worker.

Since we're advancing the last ingest time after each attachment, and known records can be sandwiched between unknown ones, we could end up in a scenario where we advance the last ingest time past an unknown record, but get terminated before we can save that unknown record's ID at the end. At that point, we won't see the record again, until the next time it's changed in Remote Settings.

(We saw a version of this in Firefox Sync, where it "missed" changes because it advanced the "last sync" timestamp out of lockstep with updating the database).

One way we could work around that—and still store the unknown records in bulk—is to keep an SQLite transaction open for the entire ingestion. But I think that's tricky; we'd need to hold the transaction open while we do HTTP, and we'd lose progress on the entire ingestion if we were interrupted halfway through.

We considered using SQLite's [built-in `json_patch` function](https://www.sqlite.org/json1.html#jpatch) to update the `UNKNOWN_RECORDS_META_KEY`, or a separate table. Serde has [excellent performance](https://github.com/serde-rs/json-benchmark), though, on par with SQLite's JSON implementation, and it's a bit easier to work with Serde. A separate table could be a good option to explore if we find ourselves accumulating large numbers of unknown record IDs.

Does that track for you?

Reply:
Of course, interruptions! I notice that `SuggestStoreInner::interrupt()` would only interrupt the reader, looks like the writer can also be interrupted by other means.

Yup, let's leave it as is since there are only a dozen of unknown records in production. We can revisit this later if needed.

Reply:
Sounds good! Yeah, between crashes, the user quitting the app, and the OS stopping background work—especially on mobile—there are lots of ways of being interrupted that we can't control.